core/src/main/java/org/apache/iceberg/IncrementalFileCleanup.java

```diff
@@ -0,0 +1,349 @@
+
+                      boolean isFromAncestor = ancestorIds.contains(manifest.snapshotId());
+                      boolean isFromExpiringSnapshot = expiredIds.contains(manifest.snapshotId());
+
+                      if (isFromAncestor && manifest.hasDeletedFiles()) {
+                        // Only delete data files that were deleted in by an expired snapshot if
+                        // that
+                        // snapshot is an ancestor of the current table state. Otherwise, a snapshot
+                        // that
+                        // deleted files and was rolled back will delete files that could be in the
+                        // current
+                        // table state.
+                        manifestsToScan.add(manifest.copy());
+                      }
+
+                      if (!isFromAncestor && isFromExpiringSnapshot && manifest.hasAddedFiles()) {
+                        // Because the manifest was written by a snapshot that is not an ancestor of
+                        // the
+                        // current table state, the files added in this manifest can be removed. The
+                        // extra
+                        // check whether the manifest was written by a known snapshot that was
+                        // expired in
+                        // this commit ensures that the full ancestor list between when the snapshot
+                        // was
+                        // written and this expiration is known and there is no missing history. If
+                        // history
+                        // were missing, then the snapshot could be an ancestor of the table state
+                        // but the
+                        // ancestor ID set would not contain it and this would be unsafe.
+                        manifestsToRevert.add(manifest.copy());
+                      }
+                    }
+                  }
+                } catch (IOException e) {
+                  throw new RuntimeIOException(
+                      e, "Failed to close manifest list: %s", snapshot.manifestListLocation());
+                }
+
+                // add the manifest list to the delete set, if present
+                if (snapshot.manifestListLocation() != null) {
+                  manifestListsToDelete.add(snapshot.manifestListLocation());
+                }
+              }
+            });
+
+    deleteDataFiles(manifestsToScan, manifestsToRevert, validIds);
+    LOG.warn("Manifests to delete: {}", Joiner.on(", ").join(manifestsToDelete));
+    LOG.warn("Manifests Lists to delete: {}", Joiner.on(", ").join(manifestListsToDelete));
+    deleteMetadataFiles(
+        manifestsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest", path, exc));
+    deleteMetadataFiles(
+        manifestListsToDelete,
+        (path, exc) -> LOG.warn("Delete failed for {}: {}", "manifest list", path, exc));
```

Comment:
Since you're already parameterizing the warning, why not just pass `"manifest", "manifest list", or "data file" into the function, rather than passing a lambda? I think that would be cleaner and would result in the same thing.

Reply:
Updated, see https://github.com/apache/iceberg/pull/5669#discussion_r990676361 for why passing in a lambda with the logging logic at runtime